### PR TITLE
Introduce a command-line interface

### LINKED ISSUE #671: feat: be able to apply gradience themes from console
### Feature description

this feature would help to be able to use gradience in scripts.
something like "flatpak run *gradience-flatpak-address* gradience apply theme.json -gtk3 -gtk4 -flatpakgtk3 -flatpakgtk4" would be immensely helpful :)

### Additional Context

_No response_

### Did you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/GradienceTeam/Gradience/blob/main/CODE_OF_CONDUCT.md).

### PR DESCRIPTION
**Pull request by tfuxu**
# Description

This PR introduces a CLI (command-line interface) for Gradience. ~~It is currently highly work-in-progress, as I still need to write functions for the majority of commands, and some of the options doesn't yet have any help information.~~ We are now at reviewing process, so in following days this PR should be merged into upstream.

### CLI to-do list:
- [x] Add essential set of commands (presets, apply, new, download, monet),
- [x] Create logic code for commands:
   - [x] `presets`
   - [x] `favorites`
   - [x] `import`
   - [x] `apply`
   - [ ] `new` (will be worked on later)
   - [x] `download`
   - [x] `monet`
   - [x] `flatpak-overrides`
   - [x] `help`
- [x] Write help information for every option,
- [x] Support for local builds,
- [x] Flatpak support

Fixes #671

## Type of change

- [ ] Bugfix (Change which fixes a issue)
- [x] New feature (Change which adds new functionality)
- [ ] Enhancement (Change which slightly improves existing code)
- [x] Breaking change (This change will introduce incompatibility with existing functionality)

## Changelog 

- Added new command-line interface,
- Moved many functions used for preset manipulation to `backend/` modules,
- Rewritten `models/preset` module to make it more readable,
- Converted some backend modules to classes,
- Made changes to log messages structure

## Testing

- [x] I have tested my changes and verified that they work as expected 

#### Currently known working build types: 
- [x] Local
- [ ] Global
- [x] Flatpak
- [ ] COPR
- [ ] AUR

#### Notes:
To run CLI in Flatpak builds, use this command: `flatpak run --command=gradience-cli com.github.GradienceTeam.Gradience.Devel`

For local builds there is a small guide on how to get CLI working:
1. Rebuild Gradience using either `local.sh` or `Makefile`,
4. Run CLI with this command: `./local_cli.sh [command]`

### CODE REVIEW

File: gradience/backend/theming/monet.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+logging = Logger()
+
+
+class Monet:
+    def __init__(self):
+        self.palette = None
+
+    def generate_from_image(self, image_path: str) -> dict:
+        #TODO: Test SVG support? I don't know what's that gradience_bg.png /
```
Comment (0xMRTT) [inline gradience/backend/theming/monet.py:37]: It's because the lib doesn't support svg so I choosed to convert SVG into PNG.

**Comment by 0xMRTT** [comment]
Thanks again @tfuxu for your contribution, if you need help for refactoring or things like that, i'll be glad to help. 

Also, since that you refactored the repo structure, it'll be nice to be able to have multiple preset open at the same time or when a preset is opening by the user instead of erasing the current one, create a new window with the new preset. (@daudix-UFO will be happy 🚀 )

**Comment by daudix-UFO** [comment]
Yep, i would like to have an option to open new gradience window, this would be really helpful when creating themes with multiple variants

File: gradience/backend/theming/monet.py (hunk 1 above)
Comment (tfuxu) [reply gradience/backend/theming/monet.py:37]: Ooh, okay 😆️ I thought that the `renderPM.drawToFile` function was accessing some file named `gradience_bg.png` to use it as a base(?) for a new image.

**Comment by tfuxu** [comment]
> Thanks again @tfuxu for your contribution, if you need help for refactoring or things like that, i'll be glad to help.

I currently need to move functions for accessing a list of installed and favorited presets from frontend to backend, so if you want, you can try to move them (put them in `theming/preset_utils` module) while I'm adding logic to other commands.

> Also, since that you refactored the repo structure, it'll be nice to be able to have multiple preset open at the same time or when a preset is opening by the user instead of erasing the current one, create a new window with the new preset. (daudix-UFO will be :rocket: )

Sure, should be relatively easy to add, but I think that this exceeds the point of this PR, so maybe in a later date I could look into that.

**Comment by 0xMRTT** [comment]
> > Thanks again @tfuxu for your contribution, if you need help for refactoring or things like that, i'll be glad to help.
> 
> I currently need to move functions for accessing a list of installed and favorited presets from frontend to backend, so if you want, you can try to move them (put them in `theming/preset_utils` module) while I'm adding logic to other commands.
> 
> > Also, since that you refactored the repo structure, it'll be nice to be able to have multiple preset open at the same time or when a preset is opening by the user instead of erasing the current one, create a new window with the new preset. (daudix-UFO will be 🚀 )
> 
> Sure, should be relatively easy to add, but I think that this exceeds the point of this PR, so maybe in a later date I could look into that.

Ofc

**Comment by tfuxu** [comment]
@0xMRTT 
Are you going to port these functions? Just asking, because I did most of the logic already and wanted to finish doing it for list commands too.

**Comment by niksingh710** [comment]
Will the be added in aur as gradience-cli ? for testing??

**Comment by tfuxu** [comment]
> Will the be added in aur as gradience-cli ? for testing??

No, you won't be able to install CLI as a standalone package, because it is tied directly to Gradience code. The only way to get CLI will be to download a full Gradience package.

**Comment by niksingh710** [comment]
> > Will the be added in aur as gradience-cli ? for testing??
> 
> No, you won't be able to install CLI as a standalone package, because it is tied directly to Gradience code. The only way to get CLI will be to download a full Gradience package.

I installed Gradience from flatpak in arch but yet unable to figure out how can i use it's cli part to apply theme.

**Comment by niksingh710** [comment]
> > Will the be added in aur as gradience-cli ? for testing??
> 
> No, you won't be able to install CLI as a standalone package, because it is tied directly to Gradience code. The only way to get CLI will be to download a full Gradience package.

how can i access this cli from flatpak?

**Comment by tfuxu** [comment]
On Flatpak you'll need to add `--command` argument before the URI of the program.
Example:
```
flatpak run --command=gradience-cli com.github.GradienceTeam.Gradience --help
```

**Comment by niksingh710** [comment]
> `flatpak run --command=gradience-cli com.github.GradienceTeam.Gradience --help`

This is the error what am facing.
`bwrap: execvp gradience-cli: No such file or directory`

**Comment by tfuxu** [comment]
What did you use to build flatpak package for Gradience?

**Comment by niksingh710** [comment]
`flatpak install flathub com.github.GradienceTeam.Gradience`

i used this command.

**Comment by tfuxu** [comment]
Well, you won't get really far with that one, beacuse CLI isn't yet available in any official Gradience release, this PR is however going to add it. As I mentioned before, CLI is directly tied to Gradience code, and because of that if you want a CLI now, you need to build Gradience from source using my fork.

**Comment by tfuxu** [comment]
Okay, I think this PR is ready for review process

**Comment by tfuxu** [comment]
@daudix-UFO Could you please check that all help messages are written correctly? I mean if there isn't any spelling and syntactical mistakes that I might have made.

**Comment by daudix-UFO** [comment]
Not a native American but will try

**Comment by daudix-UFO** [comment]
Uh oh

```
../meson.build:1:0: ERROR: Unknown options: "backend-gtk4"

A full log can be found at /run/build/libportal/_flatpak_build/meson-logs/meson-log.txt
```

**Comment by tfuxu** [comment]
> Uh oh
> 
> Building using GNOME Builder with devel profile
> 
> ```
> ../meson.build:1:0: ERROR: Unknown options: "backend-gtk4"
> 
> A full log can be found at /run/build/libportal/_flatpak_build/meson-logs/meson-log.txt
> ```

That's weird, it should just compile without any problems. I suspect Gnome Builder is a culprit here, maybe it doesn't update dependencies when compiling? For now use `flatpak-builder`, and I'll try to check how to fix this.

**Comment by daudix-UFO** [comment]
Changing `-Dbackend-gtk4=enabled` to `-Dbackends=gtk4` in devel manifest fixes the issue

**Comment by tfuxu** [comment]
Because somehow you have `libportal` version 0.6 when compiling Devel, but you should get it from `main` branch.

**Comment by daudix-UFO** [comment]
that's odd, anyway gradience is installed

**Comment by daudix-UFO** [comment]
not sure if we need those lines as those arguments are explained anyway

image

Instead, we can inspire from other help dialogs, e.g gnome-terminal

```
Usage:
  gnome-terminal [OPTION?] [-- COMMAND ?]

Help Options:
  -h, --help                      Show help options
  --help-all                      Show all help options
  --help-gtk                      Show GTK+ Options
  --help-terminal                 Show terminal options
  --help-window-options           Show per-window options
  --help-terminal-options         Show per-terminal options

Application Options:
  --load-config=FILE              Load a terminal configuration file
  --no-environment                Do not pass the environment
  --preferences                   Show preferences window
  -p, --print-environment         Print environment variables to interact with the terminal
  -v, --verbose                   Increase diagnostic verbosity
  -q, --quiet                     Suppress output
  --display=DISPLAY               X display to use
```

**Comment by daudix-UFO** [comment]
how I can download a preset? for some reason it does not work for me

```
❯ flatpak run --command=gradience-cli com.github.GradienceTeam.Gradience.Devel download -n tokyonight-dark
[Gradience] ERROR: No presets found with text: tokyonight-dark
```

**Comment by tfuxu** [comment]
You need to type display name

**Comment by daudix-UFO** [comment]
```
❯ flatpak run --command=gradience-cli com.github.GradienceTeam.Gradience.Devel download --preset-name "Tokyo Night Dark"
[Gradience] ERROR: No presets found with text: Tokyo Night Dark                                                         
```

**Comment by tfuxu** [comment]
> 

Looks like a bug. Well, I'll check what's wrong here, and I'll try to fix that.

**Comment by daudix-UFO** [comment]
`apply` command acts oddly too, looks like flatpak does not have access to home folder

```
flatpak run --env=LANG=en --command=gradience-cli com.github.GradienceTeam.Gradience.Devel import -p ~/cobalt-dark.json

(process:2): Gtk-WARNING **: 20:02:45.321: Locale not supported by C library.
	Using the fallback 'C' locale.
[Gradience] [1;36mINFO:[0m Importing preset: cobalt-dark.json
Traceback (most recent call last):
  File "/app/bin/gradience-cli", line 361, in <module>
    cli = CLI()
  File "/app/bin/gradience-cli", line 117, in __init__
    self.__parse_args()
  File "/app/bin/gradience-cli", line 138, in __parse_args
    self.import_preset(args)
  File "/app/bin/gradience-cli", line 226, in import_preset
    shutil.copy(
  File "/usr/lib/python3.10/shutil.py", line 417, in copy
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/usr/lib/python3.10/shutil.py", line 254, in copyfile
    with open(src, 'rb') as fsrc:
FileNotFoundError: [Errno 2] No such file or directory: '/var/home/daudix/cobalt-dark.json'
```

**Comment by tfuxu** [comment]
That's because you try to import preset from `/var/home/daudix` which obviously doesn't exist. You need to put absolute path.

**Comment by daudix-UFO** [comment]
hmm, strange

**Comment by daudix-UFO** [comment]
After allowing flatpak the access to home folder the output changes

```
flatpak run --env=LANG=en --command=gradience-cli com.github.GradienceTeam.Gradience.Devel import -p ~/cobalt-dark.json

Traceback (most recent call last):
  File "/app/bin/gradience-cli", line 48, in <module>
    from gradience.backend.theming.monet import Monet
  File "/var/home/daudix/Проекты/Gradience Team/Gradience/gradience/backend/theming/monet.py", line 25, in <module>
    from gradience.backend.models.preset import Preset
  File "/var/home/daudix/Проекты/Gradience Team/Gradience/gradience/backend/models/preset.py", line 25, in <module>
    from gradience.backend.logger import Logger
  File "/var/home/daudix/Проекты/Gradience Team/Gradience/gradience/backend/logger.py", line 21, in <module>
    from gradience.backend.constants import build_type
ModuleNotFoundError: No module named 'gradience.backend.constants'
```

**Comment by tfuxu** [comment]
Well, that's a bummer. I've checked `import` command now on Flatpak, and you actually need to give permissions to get it to work.

**Comment by tfuxu** [comment]
I think I have an idea how to fix this.
Maybe I've could add a function which would execute if it was detected that CLI is running in sandbox, that would ask user from what directory he wants to import a preset. It will then temporary add a permission in flatpak overrides which will give Gradience an access to that directory, and after importing is done, it will remove this permission. If it will work, I'll could implement this also in `monet` command to get it also working.

**Comment by daudix-UFO** [comment]
If only libportal were working from cli

**Comment by tfuxu** [comment]
Fixed bug from https://github.com/GradienceTeam/Gradience/pull/675#issuecomment-1356836153

**Comment by tfuxu** [comment]
Okay, I was able to hack into functions in `backend/flatpak_overrides` module and I've made a function that gives read/write access to a directory specified by a user. The only problem is that Flatpak (of course) doesn't dynamically load permission changes into running sandboxed application, so user will have to type two different commands to get it working 😒️

**Comment by daudix-UFO** [comment]
OK, it's not a big deal to enable one checkbox in Flatseal, or we can add access to downloads folder and say user that you need to put preset there?

**Comment by tfuxu** [comment]
> we can add access to downloads folder and say user that you need to put preset there?

That might be an idea to consider

**Comment by tfuxu** [comment]
Okay, another idea.
I'll do a mishmash of both mine and yours idea, so we will allow Gradience to access Downloads folder by default, but I'll use my function to create a command in CLI that will allow users to choose which additional directories they want to allow Gradience to read.

**Comment by tfuxu** [comment]
This is how it will look as a command:
```
usage: gradience-cli access-file [-h] [-l] [-a PATH] [-d PATH]

options:
  -h, --help            show this help message and exit
  -l, --list            list allowed directories and files
  -a PATH, --allow PATH
                        allow Gradience to access this file or directory
  -d PATH, --disallow PATH
                        disallow Gradience access to this file or directory
```

**Comment by tfuxu** [comment]
Okay, I've tested new commits on Flatpak and now everything works as expected.

**Review by daudix-UFO — APPROVED** [review]

**Comment by tfuxu** [comment]
Alright, @daudix-UFO everything tested? Can we merge?

**Comment by daudix-UFO** [comment]
@tfuxu I think we can! :rocket:

**Comment by niksingh710** [comment]
As it is merged now.
Is it accessible in flatpak build or have to wait for the release?
and will it be available in Aur?

**Comment by daudix-UFO** [comment]
> As it is merged now. Is it accessible in flatpak build or have to wait for the release? and will it be available in Aur?

You have to wait for the release, yes it will be available in aur

**Pull request merged**